test/Microsoft.ML.Tests/CalibratedModelParametersTests.cs

```diff
@@ -0,0 +1,163 @@
+
+            var modelAndSchemaPath = GetOutputPath("TestParameterMixingCalibratedModelParametersLoading.zip");
+            ML.Model.Save(model, data.Schema, modelAndSchemaPath);
+
+            var loadedModel = ML.Model.Load(modelAndSchemaPath, out var schema);
+            var castedModel = loadedModel as BinaryPredictionTransformer<CalibratedModelParametersBase<LinearBinaryModelParameters, PlattCalibrator>>;
+
+            Assert.NotNull(castedModel);
+
+            Type expectedInternalType = typeof(ParameterMixingCalibratedModelParameters<LinearBinaryModelParameters, PlattCalibrator>);
+            Assert.Equal(expectedInternalType, castedModel.Model.GetType());
+            Assert.Equal(model.Model.GetType(), castedModel.Model.GetType());
+            Done();
+        }
+
+        [Fact]
+        public void TestValueMapperCalibratedModelParametersLoading()
+        {
+            var data = GetDenseDataset();
+
+            var model = ML.BinaryClassification.Trainers.Gam(
+                new GamBinaryTrainer.Options { NumberOfThreads = 1}).Fit(data);
+
+            var modelAndSchemaPath = GetOutputPath("TestValueMapperCalibratedModelParametersLoading.zip");
+            ML.Model.Save(model, data.Schema, modelAndSchemaPath);
+
+            var loadedModel = ML.Model.Load(modelAndSchemaPath, out var schema);
+            var castedModel = loadedModel as BinaryPredictionTransformer<CalibratedModelParametersBase<GamBinaryModelParameters, PlattCalibrator>>;
+
+            Assert.NotNull(castedModel);
+
+            Type expectedInternalType = typeof(ValueMapperCalibratedModelParameters<GamBinaryModelParameters, PlattCalibrator>);
+            Assert.Equal(expectedInternalType, castedModel.Model.GetType());
+            Assert.Equal(model.Model.GetType(), castedModel.Model.GetType());
+            Done();
+        }
+
+
+        [Fact]
+        public void TestFeatureWeightsCalibratedModelParametersLoading()
+        {
+            var data = GetDenseDataset();
+
+            var model = ML.BinaryClassification.Trainers.FastTree(
+                new FastTreeBinaryTrainer.Options { NumberOfThreads = 1}).Fit(data);
+
+            var modelAndSchemaPath = GetOutputPath("TestFeatureWeightsCalibratedModelParametersLoading.zip");
+            ML.Model.Save(model, data.Schema, modelAndSchemaPath);
+
+            var loadedModel = ML.Model.Load(modelAndSchemaPath, out var schema);
+            var castedModel = loadedModel as BinaryPredictionTransformer<CalibratedModelParametersBase<FastTreeBinaryModelParameters, PlattCalibrator>>;
+
+            Assert.NotNull(castedModel);
+
+            Type expectedInternalType = typeof(FeatureWeightsCalibratedModelParameters<FastTreeBinaryModelParameters, PlattCalibrator>);
+            Assert.Equal(expectedInternalType, castedModel.Model.GetType());
+            Assert.Equal(model.Model.GetType(), castedModel.Model.GetType());
+            Done();
+        }
+
```

Comment:
@yaeldekel I still need to add a test for `SchemaBindableCalibratedModelParameters` similar to the others. I haven't found how to make a model that actually uses that class. I am working on this, and I hope to find one ASAP. But if you happen to know how to do this, I would greatly appreciate your advice. Thanks.

Reply:
So I've been trying to find a way to create a model that actually uses SBCMP, but so far I've been unsuccessful. If someone could tell me how to do it, it would be fantastic.

Still, I've noticed something quiet odd about how SBCMP is used in the existing codebase, and I would love if someone could clarify this to me. (perhaps, @eerhardt @yaeldekel  ? ).

If I look for the references of SBCMP outside the `Calibrator.cs` file, there are only four references. **They're all on the `Create` methods of the following classes, where the method returns a new SBMCP object**:

1. [LinearBinaryModelParameters](https://github.com/dotnet/machinelearning/blob/d531ea801a34a3018b8fa2a2f352902eb703cd25/src/Microsoft.ML.StandardTrainers/Standard/LinearModelParameters.cs#L485)

2. [GamBinaryModelParameters](https://github.com/dotnet/machinelearning/blob/d531ea801a34a3018b8fa2a2f352902eb703cd25/src/Microsoft.ML.FastTree/GamClassification.cs#L244)

3. [FastTreeBinaryModelParameters](https://github.com/dotnet/machinelearning/blob/d531ea801a34a3018b8fa2a2f352902eb703cd25/src/Microsoft.ML.FastTree/FastTreeClassification.cs#L97)

4. [FastForestBinaryModelParameters](https://github.com/dotnet/machinelearning/blob/d531ea801a34a3018b8fa2a2f352902eb703cd25/src/Microsoft.ML.FastTree/RandomForestClassification.cs#L112)

So below I write the 3 main questions that I've been asking myself while looking at all of the above references.

**If I get a working example of a model that actually uses SBCMP I might be able to answer myself question 2 and 3, but I might still want to discuss question 1, since it seems to be present in several different classes...**:

### Question 1: Is it right that these create methods aren't being called when loading?
As a matter of fact, when loading those 4 parameter classes I've listed, they don't even call their Create(env, ctx) method, but rather they call their constructor overload with (env, ctx) parameters. This situation, of ignoring the Create method, and going directly to the constructor, is something I've noticed in different classes, and I've mentioned it inside other comments of this PR and in issue #4292 point 1) of the "Problems" subheader.

**So I do wonder if I am missing something important, and maybe this is how it is supposed to work (for these classes, and also the other classes where I've seen this behavior)??**.

Notice that these 4 create methods I've listed have "0 references" when checking it with VisualStudio's CodeLense tool, so it seems to me that they are not used directly anywhere in the code... and so I am unable to find a working example of where in the codebase a SBCMP object gets actually instatiated.

### Question 2: Why are these create methods returning a SBCMP object? Are those BinaryModelParameters supposed to be loaded as SBCMP objects?
_If question 1 has a negative answer (i.e. if those create methods are supposed to be called when loading the BinaryModelParameters classes) then I have this question._

So far I've been assuming that if a class X has a 'Create(env, ctx)' method, it's because it is going to be called when loading an X object. For example, the create method of BinaryPredictionTransformer loads and returns a BinaryPredictionTransformer object. So, when loading a `LinearBinaryModelParameter` am I supposed to expect a SBCMP object?.


### Question 3: Is SBCMP supposed to work differently from the other CMPs classes?
_If question 2 has a positive answer (i.e. those create methods are supposed to be called and return a SBCMP object) then I have this question._

I don't really know how SBCMP is supposed to work, or be used, but I assume that it should be similar to the other classes that also inherit from `CalibratedModelParametersBase<,>`.

So it is also strange to me that those other classes are instantiated when fitting/training _trainer classes_ (e.g. PMCMP on [ScaLogisticRegressionBinaryTrainer](https://github.com/dotnet/machinelearning/blob/46ede664e8cff585c450a7005c7069b64df5a6f1/src/Microsoft.ML.StandardTrainers/Standard/SdcaBinary.cs#L1610) class, VMCMP on [GamBinaryTrainer ](https://github.com/dotnet/machinelearning/blob/d531ea801a34a3018b8fa2a2f352902eb703cd25/src/Microsoft.ML.FastTree/GamClassification.cs#L142)class, and FWCMP on [FastTreeBinaryTrainer](https://github.com/dotnet/machinelearning/blob/d531ea801a34a3018b8fa2a2f352902eb703cd25/src/Microsoft.ML.FastTree/FastTreeClassification.cs#L212)) whereas SBCMP only have references inside _parameter_ classes. 

Notice that the first 3 parameter classes I've listed are actually used as the predictors (submodels) of PMCMP, VMCMP and FWCMP respectively inside the tests I've added in this PR... making this whole thing somewhat more confusing to me, because if question 2 has an affirmative answer, then it would mean that SBCMP could be used as the predictor of PMCMP/VMCMP/FWCMP, right? Which doesn't make sense to me so far...

Reply:
Thanks for the thorough investigation! 

Answer 1: the reason the create method isn't being called, as you mentioned in the other comment, is because the `ComponentCatalog` is searching first for a constructor, and then for a `Create` method. Since the constructor and the `Create` methods both have the signature that `ComponentCatalog` is looking for, it is ignoring the `Create` method. I think this duplication is happening because a long time ago, and before the official release of ML.NET, `ComponentCatalog` only searched for public constructors/methods, so the constructor (which is typically private) was not found, and the `Create` method was called. Since we tried to internalize as many APIs as possible, and these `Create` methods didn't need to be part of the public API, `ComponentCatalog` was modified to search for internal and private members as well. From looking at the four `Create` methods you mentioned, this is what they do before actually calling the constructor:
```
            Contracts.CheckValue(env, nameof(env));
            env.CheckValue(ctx, nameof(ctx));
            ctx.CheckAtModel(GetVersionInfo());
```
I think these checks are worth keeping, but not sure how we should solve this - we can either change the order that `ComponentCatalog` searches to look for the `Create` method first, or change the signature of the constructor so that `ComponentCatalog` doesn't find it (change the order of the parameters?). @eerhardt, what do you think?

Answer 2: The `Create` methods that you mentioned seem very wrong to me. When a `FastTree` (or any of the others in your list) model is created from the `Fit()` method (like [here](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.FastTree/FastTreeClassification.cs#L183)), the model that is returned is not a `FastTreeBinaryModelParameters`, it is a `FeatureWeightsCalibratedModelParameters<...>`. So, when it is serialized, it is serialized as a `FeatureWeightsCalibratedModelParameters<,>`, and this is also how it is deserialized. At deserialization time the `FeatureWeightsCalibratedModelParameters<,>` constructor takes care of creating the predictor, not the other way around, so as far as I can tell, [this](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.FastTree/FastTreeClassification.cs#L95) line should always return true.
I tried looking at the history of these files as far as I could find, but I couldn't find where this code was introduced. I can only guess, that this is because years ago, the calibrators used to be a field of the predictor, so the predictors were responsible for loading them at deserialization time, and when calibrators were extracted out of the predictors, this was not handled properly for these four cases.

Answer 3: The SBCMP exists so that we can support binary predictors that are not `IValueMapper`s. I.e., predictors that use input that is not necessarily a single numeric vector of features. It was introduced when the `PipelineEnsemble` class was introduced, which is a predictor that can be created only from an entrypoint, and it doesn't have an estimator that creates it, or a public API. The only other predictor that I know of in ML.NET that is not an `IValueMapper` is `FieldAwareFactorizationMachineModelParameters`, which has its own `PredictionTransformer` and doesn't use the `BinaryPredictionTransformer`, for the exact reason that `BinaryPredictionTransformer` derives from `SingleFeaturePredictionTransformerBase`. So the bottom line is that there should never exist a `BinaryPredictionTransformer` where the type of model is `SchemaBindableCalibratedModelParameters`. You can still add a test that tests saving and loading of a `SchemaBindableCalibratedModelParameters`, you can see an example of how to create one in the test `EntryPointPipelineEnsemble`. The variable [binaryEnsembleCalibrated](https://github.com/dotnet/machinelearning/blob/master/test/Microsoft.ML.Core.Tests/UnitTests/TestEntryPoints.cs#L1582) is a `PredictorModel` with a `Predictor` of type `SchemaBindableCalibratedModelParameters`.


---
In reply to: [336265838](https://github.com/dotnet/machinelearning/pull/4306#discussion_r336265838) [](ancestors = 336265838)

Reply:
Thank you very much for all your feedback.

Since we've also discussed this offline, I'm marking this conversation as resolved. I've also written down our conclusions of the offline discussions [here](https://github.com/dotnet/machinelearning/pull/4306#issuecomment-545562898).

Thanks.

Reply:
Thank you very much for all your feedback.

Since we've also discussed this offline, I'm marking this conversation as resolved. I've also written down our conclusions of the offline discussions [here](https://github.com/dotnet/machinelearning/pull/4306#issuecomment-545562898).

Thanks.